internal/server/v2/parser.go

```diff
@@ -45,3 +47,91 @@ func splitArc(s string) ([]string, error) {
 	}
 	return parts, nil
 }
+
+func parseArc(s string) (*Arc, error) {
+	if len(s) < 2 {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "invalid arc string: %s", s)
+	}
+	arc := &Arc{}
+	if s[0:2] == "->" {
+		arc.out = true
+	} else if s[0:2] == "<-" {
+		arc.out = false
+	} else {
+		return nil, status.Errorf(
+			codes.InvalidArgument, "arc string should start with arrow, %s", s)
+	}
+	s = s[2:]
+	// No property defined; This is to fetch all the properties.
+	if len(s) == 0 {
+		return arc, nil
+	}
+	if s[0] == '[' {
+		if s[len(s)-1] != ']' {
+			return nil, status.Errorf(
+				codes.InvalidArgument, "invalid filter string: %s", s)
+		}
+		s = s[1 : len(s)-1]
+		// [prop1, prop2]
+		props := []string{}
+		pos := -1
+		for i := 0; i < len(s)-1; i++ {
+			if s[i] == ',' {
+				props = append(props, strings.TrimSpace(s[pos+1:i]))
+				pos = i
+			}
+		}
+		props = append(props, strings.TrimSpace(s[pos+1:]))
+		arc.props = props
+		return arc, nil
+	}
+	for i := 0; i < len(s); i++ {
+		if s[i] == '+' {
+			arc.prop = s[0:i]
+			arc.wildcard = "+"
+			s = s[i+1:]
+			break
+		} else if s[i] == '{' {
+			arc.prop = s[0:i]
+			s = s[i:]
+			break
+		}
+	}
+	if len(s) > 0 && s[0] == '{' {
+		if s[len(s)-1] != '}' {
+			return nil, status.Errorf(
+				codes.InvalidArgument, "invalid filter string: %s", s)
+		}
+		// {prop1:val1, prop2:val2}
+		filter := map[string]string{}
+		parts := strings.Split(s[1:len(s)-1], ",")
+		for _, p := range parts {
+			p = strings.TrimSpace(p)
+			kv := []int{}
+			for j := 0; j < len(p); j++ {
+				if p[j] == ':' {
+					kv = append(kv, j)
+				}
+			}
+			if len(kv) != 1 {
+				return nil, status.Errorf(
+					codes.InvalidArgument, "invalid filter string: %s", p)
+			}
+			k := strings.TrimSpace(p[0:kv[0]])
+			v := strings.TrimSpace(p[kv[0]+1:])
+			if len(k) == 0 || len(v) == 0 {
+				return nil, status.Errorf(
+					codes.InvalidArgument, "invalid filter string: %s", p)
+			}
```

Comment:
Can we just use  strings.Split(p, ":") here? Then check if there are 2 non-empty items.

Reply:
that's definitely much better